[mirai] add tag annotations to libra-crypto crate

stonebuddha opened the pull request:
<!--
Thank you for sending a PR. We appreciate you spending time to help improve the Libra project.

The project is undergoing daily changes. Pull Requests will be reviewed and responded to as time permits.
-->

## Motivation

This commit adds documentation to the code and helps reduce the false positives produced by MIRAI so that newly introduced problems would be easier to find. Some of the annotations make use of [tag analysis](https://github.com/facebookexperimental/MIRAI/blob/master/documentation/TagAnalysis.md) to track validity of public keys at compile time. (Note that #3846 explained the importance of public key validation.)

### Have you read the [Contributing Guidelines on pull requests](https://github.com/libra/libra/blob/master/CONTRIBUTING.md#pull-requests)?

Yes.

## Test Plan

cargo xtest libra-crypto

stonebuddha commented:
@hermanventer @huitseeker @mimoo This commit could be a first step to apply tag analysis.

hermanventer commented on crypto/crypto/src/ed25519.rs:
```diff
@@ -410,6 +431,7 @@ impl Signature for Ed25519Signature {
         message: &T,
         public_key: &Ed25519PublicKey,
     ) -> Result<()> {
+        precondition!(has_tag!(public_key, ValidPublicKeyTag)); // Public keys should be valid.
```
The comment does not add any value beyond what the annotation already provides.

mimoo replied:
I for one appreciate these kind of comments : o (I also wouldn't mind the comment being on its own line)

hermanventer commented on crypto/crypto/src/noise.rs:
```diff
@@ -645,8 +645,8 @@ impl NoiseSession {
 
-        // increment nonce
-        self.write_nonce += 1;
+        // increment nonce, which won't exceed the maximum u64 value in practice
```
This is a claim without evidence. Why?

mimoo replied:
While this is true (because the nonce is initialized to 0 and never reaches `u64::max`), if it ever becomes false then this change will introduce a bug. I would propose another change: since this function returns a `Result`, let's use a `checked_add` and return an error if the addition wraps around.

stonebuddha replied:
As suggested by @huitseeker, in order to keep this PR tag-specific, I've separated new bound checks from this PR to another PR #5540.

mimoo commented on crypto/crypto/src/ed25519.rs:
```diff
@@ -378,7 +397,9 @@ impl TryFrom<&[u8]> for Ed25519PublicKey {
         // and we have to again invoke deserialization.
-        Ed25519PublicKey::from_bytes_unchecked(bytes)
+        let key = Ed25519PublicKey::from_bytes_unchecked(bytes)?;
+        add_tag!(&key, ValidPublicKeyTag); // This key has gone through validity checks.
```
Why not using a string here instead of defining a type `ValidPublicKeyTag`?

stonebuddha replied:
The current tagging mechanism requires explicitly defined tag types. These types can be parameterized by a propagation set, which specifies the set of operations (e.g., bit-xor, subcomponents) that can propagate the tag.

I'll make the tag definition more modular.

mimoo commented on crypto/crypto/src/ed25519.rs:
```diff
@@ -76,6 +77,24 @@ impl Clone for Ed25519PrivateKey {
+/// The propagation set of ValidPublicKeyTag. An empty propagation is used to make sure that
+/// ValidPublicKeyTag can only be explicitly attached to public keys.
+#[cfg(feature = "const_generics")]
+const VALID_PUBLIC_KEY_TAG_MASK: TagPropagationSet = tag_propagation_set!();
```
not sure I understand what this really does :D

huitseeker replied:
It prevents propagating the tag (which presence marks a checked public key) through e.g. equality. 
Though in all rigor, cloning and equality should propagate `ValidPublicKeyTag`.
 
https://github.com/facebookexperimental/MIRAI/blob/master/documentation/TagAnalysis.md#declaring-tags

mimoo replied:
Ah I see, makes sense! From a quick glance I don't see a cloning here though: https://github.com/facebookexperimental/MIRAI/blob/8f0596f02b557c3e357767ee73c437d4a9bf8c5c/annotations/src/lib.rs#L33

stonebuddha replied:
It's a bit unfortunate that cloning is not yet handled properly (see this [issue](https://github.com/facebookexperimental/MIRAI/issues/611)), because essentially tags are stored as a model field of a struct, which is not transferred by cloning. I'm investigating this problem.

mimoo replied:
equal means copy?

stonebuddha replied:
Equal means equality checks. From the perspective of semantics, copying generates duplicates of a value, so it is assumed to always propagate tags.

stonebuddha replied:
Cloning now tracks tag/model fields properly (see this [PR](https://github.com/facebookexperimental/MIRAI/pull/613)).

huitseeker commented on crypto/crypto/src/ed25519.rs:
```diff
@@ -76,6 +77,24 @@ impl Clone for Ed25519PrivateKey {
 #[derive(DeserializeKey, Clone, SerializeKey)]
 pub struct Ed25519PublicKey(ed25519_dalek::PublicKey);
 
+/// A private tag type that tracks if a public key is checked to protect against invalid point
```
Would it be envisionable to put those in their own module? A [`mod`](https://doc.rust-lang.org/reference/items/modules.html) declaration would go a long way towards unifying declarations with this unified functionality.

huitseeker commented on crypto/crypto/src/hash.rs:
```diff
@@ -248,6 +248,7 @@ impl HashValue {
 
     /// Returns the `index`-th nibble.
     pub fn get_nibble(&self, index: usize) -> Nibble {
+        precondition!(index < HashValue::LENGTH);
```
Can you put this one in a separate (first) commit, please? That way we can manage tag analysis and bound checks independently (mentally, in history, in reverts, ...).

huitseeker commented on crypto/crypto/src/lib.rs:
```diff
@@ -3,6 +3,7 @@
 
 #![forbid(unsafe_code)]
 #![deny(missing_docs)]
+#![cfg_attr(mirai, allow(incomplete_features), feature(const_generics))]
```
This is bound to scare people off: `const_generics` is something we do not want anywhere near production releases, and we really want to make it really easy to check that fact.

Here's one way I think this could be achieved: 
- having a *crate* feature in the crypto crate, 
- making it non-default, and 
- making it the condition for compilation of the distinct module that 
    + introduces the tag types, 
    + and therein turns on `allow(incomplete_features)` and `#![feature(const_generics)]`.

Then ideally the `has_tag` preconditions refer to the fully-qualified name for this module.

stonebuddha replied:
It seems that `#![feature(const_generics)]` cannot be used at module level. I've tried to move the definitions of tag types to a separate tag crate, but I'm not sure how to make the tag crate only accessible to the libra-crypto crate.

huitseeker replied:
Have you tried conditional compilation of the feature gate?

`#![cfg_attr(all(feature = "mirai-analysis-on", [debug_assertions](https://doc.rust-lang.org/reference/conditional-compilation.html#debug_assertions)), feature(link_llvm_intrinsics))]`

mimoo replied:
what's wrong with const_generics?

hermanventer replied:
Nothing much, other than that is not stable. We are not using any strange features and the standard libraries are using constant generics as well. I expect that something close enough like this will become part of the stable language at some point.

huitseeker replied:
`const_generics` is not only unstable, it's [incomplete](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_feature/constant.INCOMPLETE_FEATURES.html). It's nowhere near stabilization. 

I'd urge you to consider the risk of allowing anybody to depend on the activation of such a feature (which, as it stands in the current state of the PR, is crate-wide).  In fact I'm going to formally request changes to reflect that.

bmwill replied:
Given we build and run all of libra using a stable compiler, we can have some solace in the fact that if that feature happens to get turned on somehow, that compilation should fail.

hermanventer replied:
Just to help me understand the issue better: With the code as it is in this PR, can the stable compiler compile Libra?

If so, what is wrong with the code as it is in this PR?

bmwill replied:
Given the current [CI failures](https://circleci.com/gh/libra/libra/174477?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-checks-link) it seems like there's probably some more work to be done, specifically to ensure that our CI does not try to build/test with anything that enables this feature.

hermanventer replied:
Indeed, the only way code that depends on this feature should ever be compiled is when the mirai flag is set, which is something the CI should never do.

stonebuddha replied:
The current CI failure comes from the fact that tag type definitions are moved to a separate crate in the latest commit.

Originally, the tag types were defined inside the libra-crypto crate, so there was no CI failure.

hermanventer replied:
It seems to me that doing things in the original way makes sense. I would be in favor of reverting to it.

stonebuddha replied:
Reverted back to the original way, but moved tag definitions to a separate tags module, which gets compiled only if both mirai and debug_assestions are on.

Is is still necessary to have a crate feature like `mirai-tag-analysis`, rather than use the mirai identifier directly?

huitseeker replied:
- the use of the `mirai` feature is fine,
- after having played around with the code, the current submodule for tags is great,
- I'd still love a comment above the `cfg_attr(mirai, allow(incomplete_features), ...` which states something like:
```
//! This feature gets turned on only if libra-crypto is compiled via MIRAI in a nightly build.
```

huitseeker commented on crypto/crypto/src/ed25519.rs:
```diff
@@ -76,6 +77,24 @@ impl Clone for Ed25519PrivateKey {
+type ValidPublicKeyTag = ValidPublicKeyTagKind<VALID_PUBLIC_KEY_TAG_MASK>;
+
+#[cfg(not(feature = "const_generics"))]
+struct ValidPublicKeyTag {}
```
That's an OK name, but I'd prefer something that grammatically conveys the "has-happened-before" concept, hence `ValidatedPublicKeyTag`

stonebuddha commented:
A quick reply to the first two questions in https://github.com/libra/libra/pull/5525#pullrequestreview-464472325:

> * I expect tags on trait methods are nowhere near checkable yet, is that correct?

If you mean that we annotate trait methods with contracts like "the result must have a validated tag," I think currently a possible way is to use the [design by contracts crate](https://gitlab.com/karroffel/contracts) with the `contract_trait` attribute.

> * is it possible to run cross-crate checks yet?

Yes. I'm also trying to annotate and analyze some other crates that make use of libra-crypto, such as safety-rules.

stonebuddha commented:
A reply to the last two questions in https://github.com/libra/libra/pull/5525#pullrequestreview-464472325:

> * while I have you here, how about adding tag checks to each element of the public key iterator argument  of `batch_verify`?

Currently MIRAI does not support quantification in preconditions (e.g., "all the keys in the input have the Validated tag").
We could instead use `verify!` macros, but they would probably flag many false positives.

> * if we're being super ambitious, those checks straightforwardly extend to MultiEd25519PublicKey, and both modules could share the same set of tag declarations (see my suggestion to open a module for those in PR comments)

Added annotations to MultiEd25519PublicKey.

Regarding https://github.com/libra/libra/pull/5525#discussion_r468105758:

Added conditional compilation to guard the usage of const_generics by both mirai and debug_assertions.
Is it still necessary to add an extra crate feature like `mirai-tag-analysis` as a third condition?

bors-libra commented:
:exclamation: Invalid command

<details>
<summary>Bors help and documentation</summary>
<br />

Bors is a Github App used to manage merging of PRs in order to ensure that CI is always green. It does so by maintaining a [Merge Queue](https://github.com/libra/libra/projects/2). Once a PR reaches the head of the Merge Queue it is rebased on top of the latest version of the PR's `base-branch` (generally `master`) and then triggers CI. If CI comes back green the PR is then merged into the `base-branch`. Regardless of the outcome, the next PR is the queue is then processed.

### General
- This project's Merge Queue can be found [here](https://github.com/libra/libra/projects/2).
- This project requires PRs to be [Reviewed](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/about-pull-request-reviews) and Approved before they can be queued for merging.
- Before PR's can be merged they must be configured with ["Allow edits from maintainers"](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork) enabled. This is needed for Bors to be able to update PR's in-place so that Github can properly recognize and mark them as "merged" once they've been merged into the upstream branch.
### Commands
Bors actions can be triggered by posting a comment which includes a line of the form `/<action>`.
| Command | Action | Description |
| --- | --- | --- |
| __Land__ | `land`, `merge` | attempt to land or merge a PR |
| __Canary__ | `canary`, `try` | canary a PR by performing all checks without merging |
| __Cancel__ | `cancel`, `stop` | stop an in-progress land |
| __Cherry Pick__ | `cherry-pick <target>` | cherry-pick a PR into `<target>` branch |
| __Priority__ | `priority` | set the priority level for a PR (`high`, `normal`, `low`) |
| __Help__ | `help`, `h` | show this help message |

### Options
Options for Pull Requests are configured through the application of labels.
| &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Option&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; | Description |
| --- | --- |
| ![label: bors-high-priority](https://img.shields.io/static/v1?label=&message=bors-high-priority&color=lightgrey) | Indicates that the PR is high-priority. When queued the PR will be placed at the head of the merge queue. |
| ![label: bors-low-priority](https://img.shields.io/static/v1?label=&message=bors-low-priority&color=lightgrey) | Indicates that the PR is low-priority. When queued the PR will be placed at the back of the merge queue. |
| ![label: bors-squash](https://img.shields.io/static/v1?label=&message=bors-squash&color=lightgrey) | Before merging the PR will be squashed down to a single commit, only retaining the commit message of the first commit in the PR. |

</details>

huitseeker commented:
@bors-libra /land

bors-libra commented:
:exclamation: Invalid command

<details>
<summary>Bors help and documentation</summary>
<br />

Bors is a Github App used to manage merging of PRs in order to ensure that CI is always green. It does so by maintaining a [Merge Queue](https://github.com/libra/libra/projects/2). Once a PR reaches the head of the Merge Queue it is rebased on top of the latest version of the PR's `base-branch` (generally `master`) and then triggers CI. If CI comes back green the PR is then merged into the `base-branch`. Regardless of the outcome, the next PR is the queue is then processed.

### General
- This project's Merge Queue can be found [here](https://github.com/libra/libra/projects/2).
- This project requires PRs to be [Reviewed](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/about-pull-request-reviews) and Approved before they can be queued for merging.
- Before PR's can be merged they must be configured with ["Allow edits from maintainers"](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork) enabled. This is needed for Bors to be able to update PR's in-place so that Github can properly recognize and mark them as "merged" once they've been merged into the upstream branch.
### Commands
Bors actions can be triggered by posting a comment which includes a line of the form `/<action>`.
| Command | Action | Description |
| --- | --- | --- |
| __Land__ | `land`, `merge` | attempt to land or merge a PR |
| __Canary__ | `canary`, `try` | canary a PR by performing all checks without merging |
| __Cancel__ | `cancel`, `stop` | stop an in-progress land |
| __Cherry Pick__ | `cherry-pick <target>` | cherry-pick a PR into `<target>` branch |
| __Priority__ | `priority` | set the priority level for a PR (`high`, `normal`, `low`) |
| __Help__ | `help`, `h` | show this help message |

### Options
Options for Pull Requests are configured through the application of labels.
| &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Option&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; | Description |
| --- | --- |
| ![label: bors-high-priority](https://img.shields.io/static/v1?label=&message=bors-high-priority&color=lightgrey) | Indicates that the PR is high-priority. When queued the PR will be placed at the head of the merge queue. |
| ![label: bors-low-priority](https://img.shields.io/static/v1?label=&message=bors-low-priority&color=lightgrey) | Indicates that the PR is low-priority. When queued the PR will be placed at the back of the merge queue. |
| ![label: bors-squash](https://img.shields.io/static/v1?label=&message=bors-squash&color=lightgrey) | Before merging the PR will be squashed down to a single commit, only retaining the commit message of the first commit in the PR. |

</details>

huitseeker commented:
/land

bors-libra commented:
@huitseeker :exclamation: This PR is still missing approvals, unable to queue for landing

mimoo commented on crypto/crypto/src/lib.rs:
```diff
@@ -49,3 +54,9 @@ compile_error!(
+// containing the definitions of MIRAI tag types should not get compiled in a release build.
+// The code below fails a build of the crate if mirai is on but debug_assertions is not.
+#[cfg(all(mirai, not(debug_assertions)))]
+compile_error!("MIRAI can only be used to compile the crate in a debug build!");
```
is it really the right place for this assertion?

stonebuddha replied:
It would be better if the assertion is implemented universally for the project. Any suggestions?

huitseeker commented:
/land
